### PR TITLE
Bugfix/anyurl escapes percent sign in url

### LINKED ISSUE #4458: AnyUrl escapes percent sign in URLs without TLD that are already escaped.
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic  (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

When using `AnyUrl` with a localhost URL (or any URL without a top level domain), if the provided URL already is encoded/escaped, pydantic will escape the percent sign again. When the domain has a TLD, it works perfectly fine. I've confirmed that the bug isn't present in 1.9.1. It seems to have been introduced in 1.10.0 and is still present in 1.10.1.

### Example Code

```Python
from pydantic import BaseModel, AnyUrl


class Foo(BaseModel):
    url: AnyUrl


class Bar(BaseModel):
    url: str


assert Foo(url='https://example.com/Foo%20Bar').url == 'https://example.com/Foo%20Bar'  # Passes
assert Bar(url='https://localhost/Foo%20Bar').url == 'https://localhost/Foo%20Bar'  # Passes
assert Foo(url='https://localhost/Foo%20Bar').url == 'https://localhost/Foo%20Bar'  # Fails: https://localhost/Foo%2520Bar
```

### Python, Pydantic & OS Version

```Text
$ python -c "import pydantic.utils; print(pydantic.utils.version_info())"
             pydantic version: 1.10.1
            pydantic compiled: True
                 install path: /home/afitter/.local/lib/python3.10/site-packages/pydantic
               python version: 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0]
                     platform: Linux-5.15.0-46-generic-x86_64-with-glibc2.35
     optional deps. installed: ['typing-extensions']
```

### Affected Components

- [X] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [X] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Settings Management](https://pydantic-docs.helpmanual.io/usage/settings/)
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

### PR DESCRIPTION
**Pull request by vetedde**
## Change Summary

Fix re-quote when build url. I added relevant cases to test.
Also fix mypy error, because without it I couldn't create commit

## Related issue number
Fix #4458 

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/pydantic/pydantic/blob/main/changes/README.md) for details.
  You can [skip this check](https://github.com/pydantic/hooky#change-file-checks) if the change does not need a change file.)
* [] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: pydantic/mypy.py [hunk 1]
```diff
@@ -121,7 +121,7 @@ def get_method_hook(self, fullname: str) -> Optional[Callable[[MethodContext], T
 
     def get_class_decorator_hook(self, fullname: str) -> Optional[Callable[[ClassDefContext], None]]:
         if fullname == DATACLASS_FULLNAME:
-            return dataclasses.dataclass_class_maker_callback  # type: ignore[return-value]
+            return dataclasses.dataclass_class_maker_callback
```
Comment (hramezani) [inline pydantic/mypy.py:124]: I think we don't need this change. you may have an old version of mypy in your local setup

**Comment by hramezani** [comment]
Please add tests for your change into [test_build_url_quote_plus](https://github.com/pydantic/pydantic/blob/e34ff9255f281eaccac10d0436c082626d9146c3/tests/test_networks.py#L703)

**Comment by hramezani** [comment]
Thanks @vetedde for this patch :+1: 
I left some comments. please update

File: pydantic/networks.py [hunk 2]
```diff
@@ -396,6 +396,10 @@ def apply_default_parts(cls, parts: 'Parts') -> 'Parts':
 
     @classmethod
     def quote(cls, string: str, safe: str = '') -> str:
+        pattern = r'^([\w]+|(%\d{2}))+$'
```
Comment (hramezani) [inline pydantic/networks.py:399]: please compile the pattern and then use the `match` on the compiled pattern.
Take a look at the same implementation in `datetime_parse.py`
https://github.com/pydantic/pydantic/blob/32ea8853450b50e5b963ce677ab1fff95bc2e56a/pydantic/datetime_parse.py#L30

**Comment by vetedde** [comment]
@hramezani found a problem with case "foo%20bar", when quote_plus is True. My solution returns foo%20bar, but need foo+bar. I'm thinking of adding an extra check, but it's nested check and still have to unquote at some point

```python
string_with_precent_code_re = re.compile(r'^([\w]+|(%\d{2}))+$')

if string_with_precent_code_re.match(string) is not None:
    if cls.quote_plus:
        string = unquote_plus(string)
    else:
        return string

return quote_plus(string, safe) if cls.quote_plus else quote(string, safe)
```

Maybe it's better to do it at once:

```python
string = unquote_plus(string, safe) if cls.quote_plus else unquote(string, safe)
return quote_plus(string, safe) if cls.quote_plus else quote(string, safe)
```
 or always do unquote if match %xx

What do you think?

File: pydantic/networks.py (hunk 2 above)
Comment (samuelcolvin) [reply pydantic/networks.py:399]: Can you add a comment explaining what this regex is doing?

I also think it could be simplified.

**Comment by hramezani** [comment]
@vetedde I think there are 3 possible solutions:

1. unquote string in any case and then quote it
2. unquote string if it is already quoted(checking by regex for example) and then quote it
3. revert this feature because seems it is complex and is not good performance-wise (we need to unquote and then quote)

What do you think @samuelcolvin

**Comment by samuelcolvin** [comment]
please have a look at #4469, I think it's a cleaner approach to solve this. I've also added a large number of test cases to at least lock the behaviour. I'll reply on the original issue with some more context.

Thanks so much @vetedde for looking into this, I'll leave this PR open in case it is decided that it's a better approach than #4469.

**Comment by hramezani** [comment]
I think we still will have the problem reported in https://github.com/pydantic/pydantic/issues/4468 even with this patch. which is a breaking change.

**Comment by samuelcolvin** [comment]
After some consideration and discussions, I've decided to replace this by #4470. @vetedde thanks again for your contribution.

See #4470 for a full explanation.

**Pull request closed without merging**